plugins/database/mongodb/connection_producer.go

```diff
@@ -187,3 +156,112 @@ func (c *mongoDBConnectionProducer) secretValues() map[string]interface{} {
 		c.Password: "[password]",
 	}
 }
+
+func (c *mongoDBConnectionProducer) getConnectionURL() (connURL string) {
+	connURL = dbutil.QueryHelper(c.ConnectionURL, map[string]string{
+		"username": c.Username,
+		"password": c.Password,
+	})
+	return connURL
+}
+
+func (c *mongoDBConnectionProducer) getWriteConcern() (opts *options.ClientOptions, err error) {
+	if c.WriteConcern == "" {
+		return nil, nil
+	}
+
+	input := c.WriteConcern
+
+	// Try to base64 decode the input. If successful, consider the decoded
+	// value as input.
+	inputBytes, err := base64.StdEncoding.DecodeString(input)
+	if err == nil {
+		input = string(inputBytes)
+	}
+
+	concern := &writeConcern{}
+	err = json.Unmarshal([]byte(input), concern)
+	if err != nil {
+		return nil, errwrap.Wrapf("error unmarshalling write_concern: {{err}}", err)
+	}
+
+	// Translate write concern to mongo options
+	var w writeconcern.Option
+	switch {
+	case concern.W != 0:
+		w = writeconcern.W(concern.W)
+	case concern.WMode != "":
+		w = writeconcern.WTagSet(concern.WMode)
+	default:
+		w = writeconcern.WMajority()
+	}
+
+	var j writeconcern.Option
+	switch {
+	case concern.FSync:
+		j = writeconcern.J(concern.FSync)
+	case concern.J:
+		j = writeconcern.J(concern.J)
+	default:
+		j = writeconcern.J(false)
+	}
+
+	writeConcern := writeconcern.New(
+		w,
+		j,
+		writeconcern.WTimeout(time.Duration(concern.WTimeout)*time.Millisecond))
+
+	opts = options.Client()
+	opts.SetWriteConcern(writeConcern)
+	return opts, nil
+}
+
+func (c *mongoDBConnectionProducer) getAuth() (opts *options.ClientOptions, err error) {
+	if len(c.TLSCAData) == 0 && len(c.TLSCertificateKeyData) == 0 {
+		return nil, nil // NOOP since the user/pass should be in the connection string
+	}
+	return c.getTLSAuth()
+}
+
+func (c *mongoDBConnectionProducer) getUserPassAuth() (opts *options.ClientOptions, err error) {
+	return nil, nil
+}
+
+func (c *mongoDBConnectionProducer) getTLSAuth() (opts *options.ClientOptions, err error) {
+	if len(c.TLSCAData) == 0 && len(c.TLSCertificateKeyData) == 0 {
```

Comment:
Is this condition correct? It would seem if TLSCertificateKeyData wasn't present we definitely couldn't TLS auth (i.e. `s/&&/||/ )`. But also, is TLSCAData required, or will system roots be picked up by default if not provided? I don't know. If it's the case, then the TLSCAData check can go away.

Reply:
As far as I can tell, the [tlsCAFile](https://docs.mongodb.com/manual/reference/program/mongo/#cmdoption-mongo-tlscafile) option isn't strictly tied to the [tlsCertificateKeyFile](https://docs.mongodb.com/manual/reference/program/mongo/#cmdoption-mongo-tlscertificatekeyfile). The CA file is used to validate the certificate presented by the server. I haven't tested it but as far as I can tell you could provide the CA file without doing client certificate auth.